### PR TITLE
fix: async middleware is equivalent to sync middleware - DIA-61984

### PR DESCRIPTION
**Pull request by arththebird**
## Description 

- Update async middleware and context manager to be equivalent to the sync version.
- Refactor async tests to be equivalent to sync version.
- Update pytest plugin to be consistent between sync and async version.

## Related PRs

Extracted from #105 

## Related JIRA issues

* [DIA-61984]

## Validation

New behaviour is thoroughly tested with unit and integration tests ✅ 

[DIA-61984]: https://dialoguemd.atlassian.net/browse/DIA-61984?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -88,6 +80,7 @@ def tear_down(environ):
     # reload fastapi_sqla to clear sqla deferred reflection mapping stored in Base
     importlib.reload(fastapi_sqla.models)
     importlib.reload(fastapi_sqla.sqla)
+    importlib.reload(fastapi_sqla.async_sqla)
```
Comment (arththebird) [inline tests/conftest.py:83]: Forgotten in previous refactor

File: tests/middleware/conftest.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+from unittest.mock import Mock
```
Comment (arththebird) [inline tests/middleware/conftest.py:1]: The middleware tests were refactored to improve coverage for the async version by adopting a similar pattern as the pagination tests.

File: tests/test_middleware.py [hunk 3]
```diff
@@ -1,191 +0,0 @@
-from unittest.mock import Mock, patch
```
Comment (arththebird) [inline tests/test_middleware.py:1]: The tests related to startup were moved to test_startup and the ones related to open_session moved to test_open_session.

File: tests/test_startup.py [hunk 4]
```diff
@@ -6,17 +6,6 @@
 from pytest import fixture, mark, raises
 from sqlalchemy import text
 
-pytestmark = mark.usefixtures("patch_engine_from_config", "patch_new_engine")
-
-
-@fixture(params=[True, False])
-def case_sensitive_environ(environ, request):
```
Comment (arththebird) [inline tests/test_startup.py:13]: It was causing issue with the `PYTEST_CURRENT_TEST` env var, so moved to its own test

File: fastapi_sqla/_pytest_plugin.py [hunk 5]
```diff
@@ -92,19 +96,19 @@ def sqla_modules():
...
 
 
 @fixture
-def patch_engine_from_config(request, db_url, sqla_connection, sqla_transaction):
+def patch_engine_from_config(request, sqla_connection, sqla_transaction):
     """So that all DB operations are never written to db for real."""
     from fastapi_sqla.sqla import _Session
 
-    if "dont_patch_engines" in request.keywords:
+    if "dont_patch_engines" in request.keywords:  # pragma: no cover
```
Comment (arththebird) [inline fastapi_sqla/_pytest_plugin.py:111]: We weren't actually testing it, and it was flagged as not covered after I cleaned up `test_startup`. I don't think it should be that PR responsibility to add coverage for this case in this fixture

Comment (hadrien) [reply fastapi_sqla/_pytest_plugin.py:111]: 🤔 weird that it was flagged as not covered as that marker is used in tests.

Comment (hadrien) [reply fastapi_sqla/_pytest_plugin.py:111]: (Anyway, agree it is not the responsibility of current PR to cover this)

**Review by edouardlp — APPROVED** [review]

**Pull request merged**